presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/AbstractArrowFlightClientHandler.java

```diff
@@ -0,0 +1,127 @@
+    {
+        try {
+            Optional<InputStream> trustedCertificate = Optional.empty();
+
+            Location location;
+            if (uri.isPresent()) {
+                location = new Location(uri.get());
+            }
+            else {
+                if (config.getArrowFlightServerSslEnabled() != null && !config.getArrowFlightServerSslEnabled()) {
+                    location = Location.forGrpcInsecure(config.getFlightServerName(), config.getArrowFlightPort());
+                }
+                else {
+                    location = Location.forGrpcTls(config.getFlightServerName(), config.getArrowFlightPort());
+                }
+            }
+
+            if (null == allocator) {
+                initializeAllocator();
+            }
+
+            FlightClient.Builder flightClientBuilder = FlightClient.builder(allocator, location);
+            if (config.getVerifyServer() != null && !config.getVerifyServer()) {
+                flightClientBuilder.verifyServer(false);
+            }
+            else if (config.getFlightServerSSLCertificate() != null) {
+                trustedCertificate = Optional.of(new FileInputStream(config.getFlightServerSSLCertificate()));
+                flightClientBuilder.trustedCertificates(trustedCertificate.get()).useTls();
+            }
+
+            FlightClient flightClient = flightClientBuilder.build();
+            return new ArrowFlightClient(flightClient, trustedCertificate);
+        }
+        catch (Exception e) {
+            throw new ArrowException(ARROW_FLIGHT_CLIENT_ERROR, "The flight client could not be obtained." + e.getMessage(), e);
+        }
+    }
+
+    private synchronized void initializeAllocator()
+    {
+        if (allocator == null) {
+            allocator = new RootAllocator(Long.MAX_VALUE);
+        }
+    }
+
+    protected abstract CredentialCallOption[] getCallOptions(ConnectorSession connectorSession);
+
+    public ArrowFlightConfig getConfig()
+    {
+        return config;
+    }
+
+    public ArrowFlightClient getClient(Optional<String> uri)
+    {
+        return initializeClient(uri);
+    }
+
+    public FlightInfo getFlightInfo(FlightDescriptor flightDescriptor, ConnectorSession connectorSession)
+    {
+        try (ArrowFlightClient client = getClient(Optional.empty())) {
+            CredentialCallOption[] auth = this.getCallOptions(connectorSession);
+            FlightInfo flightInfo = client.getFlightClient().getInfo(flightDescriptor, auth);
+            return flightInfo;
+        }
+        catch (Exception e) {
+            throw new ArrowException(ARROW_FLIGHT_INFO_ERROR, "The flight information could not be obtained from the flight server." + e.getMessage(), e);
+        }
+    }
+
+    public Optional<Schema> getSchema(FlightDescriptor flightDescriptor, ConnectorSession connectorSession)
+    {
+        return getFlightInfo(flightDescriptor, connectorSession).getSchemaOptional();
```

Comment:
I think this is problematic because this could involve the costly allocation of resources (like executing a query), only to discard the results and return the schema.  I think we should use the `GetSchema` RPC on the flight client.

Reply:
Updated to use `getSchema` RPC